### ISSUE TITLE
Switch to variable concepts

### ISSUE THREAD

**Issue by ericniebler** [ericniebler/stl2#379]
I'm getting tired of typing and looking at `()` everywhere. I suggest switching whole hog to variable templates. I threatened to do that in the committee and nobody objected.

*Bike Shedding!*

The change would involve renaming concept "overloads" like the cross-type `EqualityComparable`. Following the existing practice of `is_swappable`/`is_swappable_with`, the sensible choice is `EqualityComparableWith`. Another satisfactory choice would be `EqualityComparableTo`.

I suggest the suffix `With` for cross-type concepts that are inherently symmetric in nature. The suffix `To` or `Of` could be used for asymmetric cross-type concepts like `ConvertibleTo`.

*Possibly Related*

Along that vein, should we consider renaming `Sentinel` to `SentinelOf`? This reads nicer:

```c++
template <InputIterator I, SentinelOf<I> S>
void algorithm( I, S );
```

Proposed Resolution
-------------------

See [P0651](https://github.com/ericniebler/stl2/blob/master/ext/variable_concepts/variable_concepts.md) for wording.

**Comment by CaseyCarter** [comment]
Overloaded concept names that work with `With` or `To` suffix:
* `Swappable`
* `EqualityComparable`
* `StrictTotallyOrdered`

Overloaded concept names that don't:
* `Relation`
* `StrictWeakOrder`
* `IndirectInvocable`
* `IndirectRegularInvocable`
* `IndirectPredicate`

**Comment by cjdb** [comment]
> I'm getting tired of typing and looking at `()` everywhere. I suggest switching whole hog to variable templates. I threatened to do that in the committee and nobody objected.

No objections, I've always found the parens odd.

Besides some nice formatting, do function concepts give us anything extra? I used to think that they opened a scope for aliases and constant expressions, and was very disappointed when I found out this wasn't the case. </rant>

**Comment by ericniebler** [comment]
> `Relation`

I suggest dropping the `Relation<R, T>` form and redefining `Relation` as:

```c++
template <class R, class T, class U = T>
concept bool Relation() {
  return Predicate<R, T, T>() &&
    Predicate<R, U, U>() &&
    CommonReference<__add_const_ref<T>, __add_const_ref<U>>() &&
    Predicate<R,
      common_reference_t<__add_const_ref<T>, __add_const_ref<U>>,
      common_reference_t<__add_const_ref<T>, __add_const_ref<U>>>() &&
    Predicate<R, T, U>() &&
    Predicate<R, U, T>();
}
```

> `StrictWeakOrder`

Just drop `StrictWeakOrder<R, T>`.

> `IndirectInvocable`
> `IndirectRegularInvocable`
> `IndirectPredicate`

I suggest giving these numbers: `IndirectInvocable0`, `IndirectInvocable1`, and `IndirectInvocable2`. Gross, I know. Other suggestions?

I don't know how to make these take a variable number of `Readable`s and do the cross-product without the sorts of metaprogramming that range-v3 and cmcstl2 do. Concepts isn't expressive enough to achieve that on its own, I don't think.

**Comment by CaseyCarter** [comment]
> Besides some nice formatting, do function concepts give us anything extra? I used to think that they opened a scope for aliases and constant expressions, and was very disappointed when I found out this wasn't the case.

I would *love* to be able to declare e.g. `using T = rvalue_reference_t<iterator_t<SomeRange>>;` in a requires expression as if it were a function body.

The only non-`()` difference between function and variable concepts is overloading: you can declare function concepts with the same name and differing template parameter arity; variable concepts are not overloadable. You can never truly form an overload set of concepts like:
```c++
template<class T> concept bool foo() { return /*...*/; }
template<class T, class U> concept bool foo() { return /*...*/; }

template <class...Args>
  requires foo<Args...>() // Error
struct S {};
```
thanks to arcane template substitution rules that I don't really understand.

**Comment by ericniebler** [comment]
It's time to make a decision about this. It will need a paper, and I would need to start on it now. So, variable concepts, yea or nay?

> I suggest dropping the unary `Indirect(Regular)Invocable`, and numbering the others: `Indirect(Regular)Invocable1`, and `Indirect(Regular)Invocable2`. Ditto for `IndirectPredicate`. Gross, I know. Other suggestions?

`Indirect(Regular)UnaryInvocable` and `Indirect(Regular)BinaryInvocable` would be better, I think.

**Comment by W-E-Brown** [comment]
> On May 30, 2017, at 11:32 AM, Eric Niebler <notifications@github.com> wrote:
> It's time to make a decision about this. It will need a paper, and I would need to start on it now. So, variable concepts, yea or nay?
> 
FWIW, I vote yea, and emphatically so!

**Comment by timsong-cpp** [comment]
I'm tempted to vote nay because I'll have to rewrite all the things on cppreference.

Yea.

**Comment by dhouck** [comment]
Is there a reason variable concepts aren’t template-arity-overloadable like function concepts?

**Comment by CaseyCarter** [comment]
> Is there a reason variable concepts aren’t template-arity-overloadable like function concepts?

Because only functions and function templates are overloadable in C++; function concepts and variable concepts are still function / variable templates, albeit special cases thereof.

> So, variable concepts, yea or nay?

Weak yes. I would rather not have to write `()` anymore, but I don't think it's critical to make that change before publishing the TS.

**Comment by ericniebler** [comment]
> > `Relation`
>
> I suggest dropping the `Relation<R, T>` form and redefining `Relation` as:
<snip>

Yeah, that makes `Relation<F, T>` far heavier than it needs to be. Not ideal.

> > `StrictWeakOrder`
>
> Just drop `StrictWeakOrder<R, T>`.

That doesn't fly. There are too many places that use `IndirectStrictWeakOrder` with one argument type.

So @CaseyCarter, is `RelationWith<R, A, B>` really so bad? Or `StrictWeakOrderWith<R, A, B>`?

**Comment by CaseyCarter** [comment]
>  is `RelationWith<R, A, B>` really so bad? Or `StrictWeakOrderWith<R, A, B>`?

Neither is "over my dead body," and I don't have a better suggestion off the top of my head.

**Comment by timsong-cpp** [comment]
These three sound fine:

- `Swappable<A>` => "`A` is swappable.". 
- `SwappableWith<A, B>` => "`A` is swappable with `B`." 
- `Relation<R, A>` => "`R` is a relation over `A`."

But `RelationWith<R, A, B>` => "`R` is a relation over `A` with `B`"? That doesn't really make sense.

**Comment by ericniebler** [comment]
I'm open to naming suggestions. Please also make a suggestion for `StrictWeakOrder`. I'm working on the paper now.

**Comment by timsong-cpp** [comment]
Hmm, `Relation` is weird because `Relation<R, A, B>` isn't "R is a relation over `A` and `B`", it's more like "`R` is a relation over the common type of `A` and `B`". So...`CommonRelation`?

**Comment by ericniebler** [comment]
I'm being stupid. The single-argument forms `Relation<R, T>` and `StrictWeakOrder<R, T>` are used virtually nowhere in the Ranges TS. I can just drop them. (I was thrown off by the common occurrence of `IndirectRelation<R, I>` and `IndirectStrictWeakOrder<R, I>`, but those concepts are not overloaded.)

**Comment by ericniebler** [comment]
I can confirm, from CaseyCarter/cmcstl2#101, that it is safe to drop `Relation<R, T>` and `StrictWeakOrder<R, T>`.

**Comment by ericniebler** [comment]
Paper is evolving [here](https://github.com/ericniebler/stl2/blob/master/ext/variable_concepts/variable_concepts.md). Comments on direction are welcome.

**Comment by ericniebler** [comment]
Once we fix #411, there will be no uses of `IndirectBinaryPredicate`. I suggest dropping it. We will then have a few of `Unary` indirect callable concepts, but no nullary or binary. We then have the option to drop `Unary` entirely. My preference is to keep it.
